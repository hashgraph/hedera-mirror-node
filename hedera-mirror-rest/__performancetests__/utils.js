#!/usr/bin/env node
/*-
 *
 * Hedera Mirror Node
 *
 * Copyright (C) 2019 Hedera Hashgraph, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

const fs = require('fs');
const yaml = require('js-yaml');

// Load config
process.env.CONFIG_NAME = 'perfTestConfig';
process.env.CONFIG_PATH = __dirname;
const config = require('../config');

const mustLoadYaml = fileName => {
  try {
    console.log(`Loading yaml file ${fileName}`);
    return yaml.safeLoad(fs.readFileSync(fileName, 'utf8'));
  } catch (err) {
    console.log(`Failed to load yaml file ${fileName}: ${err}`);
    process.exit(1);
  }
};

module.exports = {
  config: config,
  mustLoadYaml: mustLoadYaml
};
