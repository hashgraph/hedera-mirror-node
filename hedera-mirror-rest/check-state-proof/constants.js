/*-
 *
 * Hedera Mirror Node
 *  ​
 * Copyright (C) 2019 - 2021 Hedera Hashgraph, LLC
 *  ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

'use strict';

const MAX_RECORD_LENGTH = 64 * 1024;

const MAX_TRANSACTION_LENGTH = 64 * 1024;

const SHA_384_LENGTH = 48;

const SHA_384_WITH_RSA = {
  type: 1,
  maxLength: 384,
};

const SIMPLE_SUM = 101;

module.exports = {
  MAX_RECORD_LENGTH,
  MAX_TRANSACTION_LENGTH,
  SHA_384_LENGTH,
  SHA_384_WITH_RSA,
  SIMPLE_SUM,
};
