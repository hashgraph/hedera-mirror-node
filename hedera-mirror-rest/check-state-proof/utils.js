/*-
 * ‌
 * Hedera Mirror Node
 * ​
 * Copyright (C) 2019 - 2020 Hedera Hashgraph, LLC
 * ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */
'uses strict';

// external libraries
const fs = require('fs');
const fetch = require('node-fetch');
const AbortController = require('abort-controller');

const base64StringToBuffer = (base64String) => {
  return Buffer.from(base64String, 'base64');
};

const replaceSpecialCharsWithUnderScores = (stringToFormat) => {
  return stringToFormat.replace(/[.@\-]/g, '_');
};

const makeStateProofDir = (transactionId, stateProofJson) => {
  const newDirPath = replaceSpecialCharsWithUnderScores(transactionId);
  fs.mkdirSync(newDirPath, {recursive: true});
  fs.writeFileSync(`${newDirPath}/apiResponse.json`, JSON.stringify(stateProofJson));
  console.log(`Supporting files and API response for the state proof will be stored in the directory ${newDirPath}`);
  return newDirPath;
};

const storeFile = (data, file, ext) => {
  const newFilePath = `${replaceSpecialCharsWithUnderScores(file)}.${ext}`;
  fs.writeFileSync(`${newFilePath}`, data, (err) => {
    if (err) throw err;
  });
};

const getAPIResponse = async (url) => {
  const controller = new AbortController();
  const timeout = setTimeout(
    () => {
      controller.abort();
    },
    60 * 1000 // in ms
  );

  console.log(`Requesting state proof files from ${url}...`);
  return await fetch(url, {signal: controller.signal})
    .then(async (response) => {
      if (!response.ok) {
        throw Error(response.statusText);
      }
      return response.json();
    })
    .catch((error) => {
      throw `Error fetching ${url}: ${error}`;
    })
    .finally(() => {
      clearTimeout(timeout);
    });
};

const readJSONFile = (filePath) => {
  const rawData = fs.readFileSync(filePath);
  return JSON.parse(rawData);
};

module.exports = {
  base64StringToBuffer,
  getAPIResponse,
  makeStateProofDir,
  readJSONFile,
  storeFile,
  replaceSpecialCharsWithUnderScores,
};
