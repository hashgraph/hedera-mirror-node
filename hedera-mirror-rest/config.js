/*-
 * ‌
 * Hedera Mirror Node
 *
 * Copyright (C) 2019 - 2020 Hedera Hashgraph, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */

'use strict';

const extend = require('extend');
const fs = require('fs');
const yaml = require('js-yaml');
const path = require('path');
const {InvalidConfigError} = require('./errors/invalidConfigError');
const {cloudProviders, networks, defaultBucketNames} = require('./constants');

let configName = 'application';
if (process.env.CONFIG_NAME) {
  configName = process.env.CONFIG_NAME;
}

const config = {};
let loaded = false;

function load(configPath) {
  if (!configPath) {
    return;
  }

  let configFile = path.join(configPath, `${configName}.yml`);
  if (fs.existsSync(configFile)) {
    loadYaml(configFile);
  }

  configFile = path.join(configPath, configName + '.yaml');
  if (fs.existsSync(configFile)) {
    loadYaml(configFile);
  }
}

function loadYaml(configFile) {
  try {
    const doc = yaml.safeLoad(fs.readFileSync(configFile, 'utf8'));
    console.log(`Loaded configuration source: ${configFile}`);
    extend(true, config, doc);
  } catch (err) {
    console.log(`Skipping configuration ${configFile}: ${err}`);
  }
}

function loadEnvironment() {
  for (const [key, value] of Object.entries(process.env)) {
    setConfigValue(key, value);
  }
}

/*
 * Sets a config property from an environment variable by converting HEDERA_MIRROR_REST_FOO_BAR to an object path
 * notation hedera.mirror.rest.foo.bar using a case insensitive search. If more than one property matches with a
 * different case, it will choose the first.
 */
function setConfigValue(propertyPath, value) {
  let current = config;
  let properties = propertyPath.toLowerCase().split('_');

  // Ignore properties that don't start with HEDERA_MIRROR_REST
  if (properties.length < 4 || properties[0] !== 'hedera' || properties[1] !== 'mirror' || properties[2] !== 'rest') {
    return;
  }

  for (let i in properties) {
    let property = properties[i];
    let found = false;

    for (let [k, v] of Object.entries(current)) {
      if (property === k.toLowerCase()) {
        if (i < properties.length - 1) {
          current = v;
          found = true;
          break;
        } else {
          current[k] = convertType(value);
          console.log(`Override config with environment variable ${propertyPath}=${value}`);
          return;
        }
      }
    }

    if (!found) {
      return;
    }
  }
}

function convertType(value) {
  let parsedValue = value;

  if (value !== null && value !== '' && !isNaN(value)) {
    parsedValue = +value;
  } else if (value === 'true' || value === 'false') {
    parsedValue = value === 'true';
  }

  return parsedValue;
}

function getConfig() {
  return config.hedera && config.hedera.mirror ? config.hedera.mirror.rest : config;
}

function parseStateProofStreamsConfig() {
  const {stateproof} = getConfig();
  if (!stateproof || !stateproof.enabled) {
    return;
  }

  const {streams: streamsConfig} = stateproof;
  if (!Object.values(networks).includes(streamsConfig.network)) {
    throw new InvalidConfigError(`unknown network ${streamsConfig.network}`);
  }

  if (!streamsConfig.bucketName) {
    streamsConfig.bucketName = defaultBucketNames[streamsConfig.network];
  }

  if (!streamsConfig.bucketName) {
    // the default for network 'OTHER' is null, throw err if it's not configured
    throw new InvalidConfigError('stateproof.streams.bucketName must be set');
  }

  if (!Object.values(cloudProviders).includes(streamsConfig.cloudProvider)) {
    throw new InvalidConfigError(`unsupported object storage service provider ${streamsConfig.cloudProvider}`);
  }
}

if (!loaded) {
  load(path.join(__dirname, 'config'));
  load(__dirname);
  load(process.env.CONFIG_PATH);
  loadEnvironment();
  parseStateProofStreamsConfig();
  loaded = true;
}

module.exports = getConfig();
