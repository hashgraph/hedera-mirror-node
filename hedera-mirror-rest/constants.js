/*-
 * ‌
 * Hedera Mirror Node
 *​
 * Copyright (C) 2019 - 2021 Hedera Hashgraph, LLC
 *​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

'use strict';

// url query filer keys
const filterKeys = {
  ACCOUNT_ID: 'account.id',
  ACCOUNT_BALANCE: 'account.balance',
  ACCOUNT_PUBLICKEY: 'account.publickey',
  ENCODING: 'encoding',
  ENTITY_PUBLICKEY: 'publickey',
  LIMIT: 'limit',
  ORDER: 'order',
  RESULT: 'result',
  SEQUENCE_NUMBER: 'sequencenumber',
  TIMESTAMP: 'timestamp',
  TOKENID: 'tokenid',
  TOKEN_ID: 'token.id',
  CREDIT_TYPE: 'type',
  TRANSACTION_TYPE: 'transactiontype',
};

// sql table columns
const entityColumns = {
  ENTITY_NUM: 'entity_num',
  ENTITY_REALM: 'entity_realm',
  ENTITY_SHARD: 'entity_shard',
  PUBLIC_KEY: 'ed25519_public_key_hex',
};

const transactionColumns = {
  TYPE: 'type',
};

const requestIdLabel = 'requestId';
const responseDataLabel = 'mirrorRestData';

const orderFilterValues = {
  ASC: 'asc',
  DESC: 'desc',
};

// topic messages filter options
const characterEncoding = {
  BASE64: 'base64',
  UTF8: 'utf-8',
};

const transactionResultFilter = {
  SUCCESS: 'success',
  FAIL: 'fail',
};

const cryptoTransferType = {
  CREDIT: 'credit',
  DEBIT: 'debit',
};

const cloudProviders = {
  S3: 'S3',
  GCP: 'GCP',
};

const defaultCloudProviderEndpoints = {
  [cloudProviders.S3]: 'https://s3.amazonaws.com',
  [cloudProviders.GCP]: 'https://storage.googleapis.com',
};

const networks = {
  DEMO: 'DEMO',
  MAINNET: 'MAINNET',
  TESTNET: 'TESTNET',
  PREVIEWNET: 'PREVIEWNET',
  OTHER: 'OTHER',
};

const defaultBucketNames = {
  [networks.DEMO]: 'hedera-demo-streams',
  [networks.MAINNET]: 'hedera-mainnet-streams',
  [networks.TESTNET]: 'hedera-stable-testnet-streams-2020-08-27',
  [networks.PREVIEWNET]: 'hedera-preview-testnet-streams',
  [networks.OTHER]: null,
};

const recordStreamPrefix = 'recordstreams/record';

module.exports = {
  characterEncoding,
  cloudProviders,
  cryptoTransferType,
  defaultBucketNames,
  defaultCloudProviderEndpoints,
  entityColumns,
  filterKeys,
  networks,
  orderFilterValues,
  recordStreamPrefix,
  requestIdLabel,
  responseDataLabel,
  transactionColumns,
  transactionResultFilter,
};
