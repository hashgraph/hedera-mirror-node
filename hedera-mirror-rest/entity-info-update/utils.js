/*-
 * ‌
 * Hedera Mirror Node
 * ​
 * Copyright (C) 2019 - 2021 Hedera Hashgraph, LLC
 * ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */

'use strict';

// external libraries
const _ = require('lodash');
const fs = require('fs');
const log4js = require('log4js');
const math = require('mathjs');

// local
const config = require('./config');

const logger = log4js.getLogger();

const constructEntity = (index, headerRow, entityRow) => {
  const entityObj = {};
  const splitEntityRow = Array.from(entityRow.split(',')).filter((x) => x != null);

  for (let i = 0; i < headerRow.length; i++) {
    entityObj[headerRow[i].trim()] = splitEntityRow[i].trim();
  }

  return entityObj;
};

const readEntityCSVFileSync = () => {
  logger.info(`Parsing csv entity file ...`);
  const entities = [];

  const data = fs.readFileSync(config.filePath, 'utf-8');

  const fileContent = data.split('\n');
  const headers = Array.from(fileContent[0].split(',')).filter((x) => x != null);

  // ensure 1st column is entity num of expected format
  if (!_.eq(headers[0], 'entity')) {
    throw Error("CSV must have a header column with 1st column being 'entity'");
  }

  for (let i = 1; i < fileContent.length; i++) {
    if (!fileContent[i]) {
      // End of file
      break;
    }

    entities.push(constructEntity(i, headers, fileContent[i]));
  }

  logger.info(`${entities.length} entities were extracted from ${config.filePath}`);

  return entities;
};

/**
 * Converts seconds since epoch (seconds nnnnnnnnn format) to  nanoseconds
 * @param {String} Seconds since epoch (seconds.nnnnnnnnn format)
 * @return {String} ns Nanoseconds since epoch
 */
const secNsToNs = (secNs, ns) => {
  return math
    .add(math.multiply(math.bignumber(secNs.toString()), math.bignumber(1e9)), math.bignumber(ns.toString()))
    .toString();
};

module.exports = {
  readEntityCSVFileSync,
  secNsToNs,
};
