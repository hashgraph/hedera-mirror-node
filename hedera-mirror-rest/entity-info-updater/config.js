/*-
 * ‌
 * Hedera Mirror Node
 * ​
 * Copyright (C) 2019 - 2021 Hedera Hashgraph, LLC
 * ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */

'use strict';

const fs = require('fs');
const log4js = require('log4js');
const path = require('path');
const yaml = require('js-yaml');

const logger = log4js.getLogger();
let config = {};
let loaded = false;

const loadYaml = () => {
  try {
    const configFile = path.join(__dirname, 'config', `application.yml`);
    logger.info(`Loaded configuration source: ${configFile}`);
    config = yaml.load(fs.readFileSync(configFile, 'utf-8'));
  } catch (err) {
    logger.info(err);
  }
};

if (!loaded) {
  loadYaml();
  loaded = true;
}

const getConfig = () => {
  return config.hedera && config.hedera.mirror ? config.hedera.mirror.entityUpdate : config;
};

module.exports = getConfig();
