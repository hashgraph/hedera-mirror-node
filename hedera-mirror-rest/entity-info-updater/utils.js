/*-
 * ‌
 * Hedera Mirror Node
 * ​
 * Copyright (C) 2019 - 2021 Hedera Hashgraph, LLC
 * ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */

'use strict';

// external libraries

const proto = require('@hashgraph/proto');
const {KeyList, PublicKey} = require('@hashgraph/sdk');
const _ = require('lodash');
const fs = require('fs');
const log4js = require('log4js');
const path = require('path');

// local
const config = require('./config');

const logger = log4js.getLogger();

// used for timestamp range
const longMinValue = -9223372036854775808n;
const longMaxValue = 9223372036854775807n;

// used for entityId conversion
const realmOffset = 2n ** 32n; // realm is followed by 32 bits entity_num
const shardOffset = 2n ** 48n; // shard is followed by 16 bits realm and 32 bits entity_num

// sample location details
const entityCachePath = process.env.ENTITY_CACHE_PATH || `${__dirname}/sample`;
const dbCacheFileName = process.env.DB_ENTITY_CACHE_FILE || `dbEntityCache.json`;
const networkCacheFileName = process.env.NETWORK_ENTITY_CACHE_FILE || `networkEntityCache.json`;

const constructEntity = (index, headerRow, entityRow) => {
  const entityObj = {};
  const splitEntityRow = Array.from(entityRow.split(',')).filter((x) => x != null);

  for (let i = 0; i < headerRow.length; i++) {
    entityObj[headerRow[i].trim()] = splitEntityRow[i].trim();
  }

  return entityObj;
};

const readEntityCSVFileSync = () => {
  logger.info(`Parsing csv entity file ...`);
  const csvStart = process.hrtime();
  const entities = [];

  const data = fs.readFileSync(config.filePath, 'utf-8');

  const fileContent = data.split('\n');
  const headers = Array.from(fileContent[0].split(',')).filter((x) => x != null);

  // ensure first column is entity num of expected format
  if (!_.eq(headers[0], 'entity')) {
    throw Error("CSV must have a header column with first column being 'entity'");
  }

  for (let i = 1; i < fileContent.length; i++) {
    if (!fileContent[i]) {
      // End of file
      break;
    }

    entities.push(constructEntity(i, headers, fileContent[i]));
  }

  const elapsedTime = process.hrtime(csvStart);
  logger.info(
    `${entities.length} entities were extracted from ${config.filePath} in ${getElapsedTimeString(elapsedTime)}`
  );

  return entities;
};

/**
 * Converts timestamp of seconds since epoch (seconds nnnnnnnnn format) to  nanoseconds
 * @param {String} Seconds since epoch (seconds.nnnnnnnnn format)
 * @return {BigInt} ns Nanoseconds since epoch
 */
const timestampToNs = (sec, ns) => {
  let finalNs = 0;

  try {
    finalNs = BigInt(sec.toString()) * BigInt(1e9) + BigInt(ns.toString());
  } catch (e) {
    logger.warn(`Error parsing sec: ${JSON.stringify(sec)}, ns: ${JSON.stringify(ns)} to timestamp`);
    throw e;
  }

  // handle the equivalent long overflow case that the java importer accommodates
  if (finalNs < longMinValue) {
    return longMinValue;
  }

  if (finalNs > longMaxValue) {
    return longMaxValue;
  }

  return finalNs;
};

const keyListToProto = (keyList) => {
  const keys = [];

  for (const key of keyList) {
    keys.push(keyToProto(key));
  }

  return {
    keys,
  };
};

const keyToProto = (key) => {
  if (key instanceof PublicKey) {
    return {
      ed25519: key.toBytes(),
    };
  }

  if (key instanceof KeyList) {
    return {
      keyList: keyListToProto(key),
    };
  }

  throw Error('Unsupported key type');
};

const getProtoAndEd25519HexFromPublicKey = (key) => {
  const protoKey = {
    ed25519: key.toBytes(),
  };
  const ed25519Hex = key.toString();

  return {protoKey, ed25519Hex};
};

const getProtoAndEd25519HexFromKeyList = (key) => {
  const protoKey = {
    keyList: keyListToProto(key),
  };

  // only keyLists of length one have an applicable ed25519Hex
  let ed25519Hex = null;
  // some complex keys are nested, ensure top level key is the applicable ed25519Hex
  if (key._keys.length === 1 && key._keys[0]._keyData !== undefined) {
    ed25519Hex = key._keys[0].toString();
  }

  return {protoKey, ed25519Hex};
};

const getBufferAndEd25519HexFromKey = (key) => {
  const {protoKey, ed25519Hex} =
    key instanceof PublicKey ? getProtoAndEd25519HexFromPublicKey(key) : getProtoAndEd25519HexFromKeyList(key);
  let protoBuffer = null;
  protoBuffer = proto.Key.encode(protoKey, protoBuffer).finish();

  return {protoBuffer, ed25519Hex};
};

const getElapsedTimeString = (elapsedTime) => {
  return `${elapsedTime[0]} s ${elapsedTime[1] / 1000000} ms`;
};

/**
 * Converts sdk AccountId string to an encoded id corresponding to the Entity. Supports 'shard.realm.num' only
 *
 * @param {AccountId|null} entityIdStr
 * @return {string|null}
 */
const getEntityId = (entityIdStr) => {
  if (_.isNull(entityIdStr)) {
    return null;
  }

  // [shard,realm,num] - entityIdStr comes from network and will always conforming to the expected format
  const parts = entityIdStr.toString().split('.');

  return (BigInt(parts[2]) + BigInt(parts[1]) * realmOffset + BigInt(parts[0]) * shardOffset).toString();
};

const getNetworkEntityCache = () => {
  const pathString = path.join(entityCachePath, networkCacheFileName);
  return fs.existsSync(pathString) ? JSON.parse(fs.readFileSync(pathString, 'utf-8')) : {};
};

const getDbEntityCache = () => {
  const pathString = path.join(entityCachePath, dbCacheFileName);
  return fs.existsSync(pathString) ? JSON.parse(fs.readFileSync(pathString, 'utf-8')) : {};
};

module.exports = {
  getBufferAndEd25519HexFromKey,
  getElapsedTimeString,
  getDbEntityCache,
  getEntityId,
  getNetworkEntityCache,
  readEntityCSVFileSync,
  timestampToNs,
};
