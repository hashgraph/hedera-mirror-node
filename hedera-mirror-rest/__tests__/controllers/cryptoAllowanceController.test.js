/*-
 * ‌
 * Hedera Mirror Node
 * ​
 * Copyright (C) 2019 - 2023 Hedera Hashgraph, LLC
 * ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */

import {getResponseLimit} from '../../config';
import * as constants from '../../constants';
import {CryptoAllowanceController} from '../../controllers';
import * as utils from '../../utils';

const {default: defaultLimit} = getResponseLimit();

const ownerIdFilter = 'owner = $1';
describe('extractCryptoAllowancesQuery', () => {
  const defaultExpected = {
    conditions: [ownerIdFilter],
    params: [1],
    order: constants.orderFilterValues.DESC,
    limit: defaultLimit,
  };

  const specs = [
    {
      name: 'limit',
      input: {
        filters: [
          {
            key: constants.filterKeys.LIMIT,
            operator: utils.opsMap.eq,
            value: 20,
          },
        ],
        accountId: 1,
      },
      expected: {
        ...defaultExpected,
        limit: 20,
      },
    },
    {
      name: 'order asc',
      input: {
        filters: [
          {
            key: constants.filterKeys.ORDER,
            operator: utils.opsMap.eq,
            value: constants.orderFilterValues.ASC,
          },
        ],
        accountId: 1,
      },
      expected: {
        ...defaultExpected,
        order: constants.orderFilterValues.ASC,
        conditions: [ownerIdFilter],
      },
    },
    {
      name: 'spender.id single',
      input: {
        filters: [
          {
            key: constants.filterKeys.SPENDER_ID,
            operator: utils.opsMap.eq,
            value: '1000',
          },
        ],
        accountId: 3,
      },
      expected: {
        ...defaultExpected,
        conditions: [ownerIdFilter, 'spender in ($2)'],
        params: [3, '1000'],
      },
    },
    {
      name: 'spender.id multiple',
      input: {
        filters: [
          {
            key: constants.filterKeys.SPENDER_ID,
            operator: utils.opsMap.eq,
            value: '1000',
          },
          {
            key: constants.filterKeys.SPENDER_ID,
            operator: utils.opsMap.eq,
            value: '1001',
          },
          {
            key: constants.filterKeys.SPENDER_ID,
            operator: utils.opsMap.eq,
            value: '1002',
          },
        ],
        accountId: 3,
      },
      expected: {
        ...defaultExpected,
        conditions: [ownerIdFilter, 'spender in ($2,$3,$4)'],
        params: [3, '1000', '1001', '1002'],
      },
    },
    {
      name: 'spender.id all allowed operators',
      input: {
        filters: [
          {
            key: constants.filterKeys.SPENDER_ID,
            operator: utils.opsMap.eq,
            value: '1000',
          },
          {
            key: constants.filterKeys.SPENDER_ID,
            operator: utils.opsMap.gt,
            value: '200',
          },
          {
            key: constants.filterKeys.SPENDER_ID,
            operator: utils.opsMap.gte,
            value: '202',
          },
          {
            key: constants.filterKeys.SPENDER_ID,
            operator: utils.opsMap.lt,
            value: '3000',
          },
          {
            key: constants.filterKeys.SPENDER_ID,
            operator: utils.opsMap.lte,
            value: '3005',
          },
        ],
        accountId: 3,
      },
      expected: {
        ...defaultExpected,
        conditions: [
          ownerIdFilter,
          'spender > $2',
          'spender >= $3',
          'spender < $4',
          'spender <= $5',
          'spender in ($6)',
        ],
        params: [3, '200', '202', '3000', '3005', '1000'],
      },
    },
  ];

  specs.forEach((spec) => {
    test(`${spec.name}`, () => {
      expect(CryptoAllowanceController.extractCryptoAllowancesQuery(spec.input.filters, spec.input.accountId)).toEqual(
        spec.expected
      );
    });
  });
});

describe('validateExtractCryptoAllowancesQuery throw', () => {
  const specs = [
    {
      name: 'spender.id ne',
      input: {
        filters: [
          {
            key: constants.filterKeys.SPENDER_ID,
            operator: utils.opsMap.ne,
            value: '1000',
          },
        ],
        accountId: 3,
      },
    },
  ];

  specs.forEach((spec) => {
    test(`${spec.name}`, () => {
      expect(() =>
        CryptoAllowanceController.extractCryptoAllowancesQuery(spec.input.filters, spec.input.accountId)
      ).toThrowErrorMatchingSnapshot();
    });
  });
});
