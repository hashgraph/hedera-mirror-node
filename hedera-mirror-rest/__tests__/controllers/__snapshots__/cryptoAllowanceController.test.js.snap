// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validateExtractCryptoAllowancesQuery throw spender.id ne 1`] = `"Not equal (ne) comparison operator is not supported for spender.id"`;
